In [2]:
# Importing necessary packages
import numpy as np 
import polars as pl

In [3]:
# Reading in the data
df = pl.read_csv(r"C:\Users\gbore\Downloads\MMA\ufc.csv")

In [4]:
# Taking a high level look at the data using head
df.head()

shape: (5, 144)
┌─────────────┬─────────────┬────────────┬────────────┬───┬───────────┬────────────┬───────┬───────┐
│ R_fighter   ┆ B_fighter   ┆ Referee    ┆ date       ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age ┆ R_age │
│ ---         ┆ ---         ┆ ---        ┆ ---        ┆   ┆ ms        ┆ bs         ┆ ---   ┆ ---   │
│ str         ┆ str         ┆ str        ┆ str        ┆   ┆ ---       ┆ ---        ┆ f64   ┆ f64   │
│             ┆             ┆            ┆            ┆   ┆ f64       ┆ f64        ┆       ┆       │
╞═════════════╪═════════════╪════════════╪════════════╪═══╪═══════════╪════════════╪═══════╪═══════╡
│ Adrian      ┆ Gustavo     ┆ Chris      ┆ 2021-03-20 ┆ … ┆ 177.8     ┆ 135.0      ┆ 31.0  ┆ 27.0  │
│ Yanez       ┆ Lopez       ┆ Tognoni    ┆            ┆   ┆           ┆            ┆       ┆       │
│ Trevin      ┆ Roman       ┆ Herb Dean  ┆ 2021-03-20 ┆ … ┆ 187.96    ┆ 185.0      ┆ 32.0  ┆ 28.0  │
│ Giles       ┆ Dolidze     ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Tai Tuivasa ┆ Harry       ┆ Herb Dean  ┆ 2021-03-20 ┆ … ┆ 190.5     ┆ 264.0      ┆ 32.0  ┆ 28.0  │
│             ┆ Hunsucker   ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Cheyanne    ┆ Montserrat  ┆ Mark Smith ┆ 2021-03-20 ┆ … ┆ 160.02    ┆ 115.0      ┆ 28.0  ┆ 25.0  │
│ Buys        ┆ Conejo      ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
│ Marion      ┆ Macy        ┆ Mark Smith ┆ 2021-03-20 ┆ … ┆ 172.72    ┆ 135.0      ┆ 29.0  ┆ 43.0  │
│ Reneau      ┆ Chiasson    ┆            ┆            ┆   ┆           ┆            ┆       ┆       │
└─────────────┴─────────────┴────────────┴────────────┴───┴───────────┴────────────┴───────┴───────┘

We looked at the first 5 rows of each of the 144 columns here.

In [5]:
# Taking a high level look at the data using tail
df.tail()

shape: (5, 144)
┌─────────────┬─────────────┬────────────┬────────────┬───┬───────────┬────────────┬───────┬───────┐
│ R_fighter   ┆ B_fighter   ┆ Referee    ┆ date       ┆ … ┆ R_Reach_c ┆ R_Weight_l ┆ B_age ┆ R_age │
│ ---         ┆ ---         ┆ ---        ┆ ---        ┆   ┆ ms        ┆ bs         ┆ ---   ┆ ---   │
│ str         ┆ str         ┆ str        ┆ str        ┆   ┆ ---       ┆ ---        ┆ f64   ┆ f64   │
│             ┆             ┆            ┆            ┆   ┆ f64       ┆ f64        ┆       ┆       │
╞═════════════╪═════════════╪════════════╪════════════╪═══╪═══════════╪════════════╪═══════╪═══════╡
│ Remco       ┆ Alberta     ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 260.0      ┆ null  ┆ null  │
│ Pardoel     ┆ Cerra Leon  ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Orlando     ┆ Robert      ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 170.0      ┆ null  ┆ null  │
│ Wiet        ┆ Lucarelli   ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Johnny      ┆ David       ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 210.0      ┆ null  ┆ null  │
│ Rhodes      ┆ Levicki     ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Patrick     ┆ Ray Wizard  ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 225.0      ┆ null  ┆ 30.0  │
│ Smith       ┆             ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
│ Scott       ┆ Sean        ┆ John       ┆ 1994-03-11 ┆ … ┆ null      ┆ 210.0      ┆ 18.0  ┆ null  │
│ Morris      ┆ Daugherty   ┆ McCarthy   ┆            ┆   ┆           ┆            ┆       ┆       │
└─────────────┴─────────────┴────────────┴────────────┴───┴───────────┴────────────┴───────┴───────┘

We looked at the last 5 rows of each of the 144 columns here.

In [8]:
df.columns

['R_fighter',
 'B_fighter',
 'Referee',
 'date',
 'location',
 'Winner',
 'title_bout',
 'weight_class',
 'B_avg_KD',
 'B_avg_opp_KD',
 'B_avg_SIG_STR_pct',
 'B_avg_opp_SIG_STR_pct',
 'B_avg_TD_pct',
 'B_avg_opp_TD_pct',
 'B_avg_SUB_ATT',
 'B_avg_opp_SUB_ATT',
 'B_avg_REV',
 'B_avg_opp_REV',
 'B_avg_SIG_STR_att',
 'B_avg_SIG_STR_landed',
 'B_avg_opp_SIG_STR_att',
 'B_avg_opp_SIG_STR_landed',
 'B_avg_TOTAL_STR_att',
 'B_avg_TOTAL_STR_landed',
 'B_avg_opp_TOTAL_STR_att',
 'B_avg_opp_TOTAL_STR_landed',
 'B_avg_TD_att',
 'B_avg_TD_landed',
 'B_avg_opp_TD_att',
 'B_avg_opp_TD_landed',
 'B_avg_HEAD_att',
 'B_avg_HEAD_landed',
 'B_avg_opp_HEAD_att',
 'B_avg_opp_HEAD_landed',
 'B_avg_BODY_att',
 'B_avg_BODY_landed',
 'B_avg_opp_BODY_att',
 'B_avg_opp_BODY_landed',
 'B_avg_LEG_att',
 'B_avg_LEG_landed',
 'B_avg_opp_LEG_att',
 'B_avg_opp_LEG_landed',
 'B_avg_DISTANCE_att',
 'B_avg_DISTANCE_landed',
 'B_avg_opp_DISTANCE_att',
 'B_avg_opp_DISTANCE_landed',
 'B_avg_CLINCH_att',
 'B_avg_CLINCH_lande

The column names are shortened but record 
- Generic information: Date, location, referee name, whether the fight was for a title and the weight class
- Fighter career statistics: Names, average knockdowns for/against, average significant strikes percentage/attempted/landed for/against, average takedown percentage/attmpted/landed for/against, attempted submissions for/against, average reversals for/against, average total strikes attempted/landed for/against, average head/body/leg strikes attempted/landed for against,

In [29]:
for x in df:
    y = x.value_counts().sort(by = 'counts', descending = True)
    print(y)

shape: (1_514, 2)
┌───────────────────────┬────────┐
│ R_fighter             ┆ counts │
│ ---                   ┆ ---    │
│ str                   ┆ u32    │
╞═══════════════════════╪════════╡
│ Jim Miller            ┆ 27     │
│ Diego Sanchez         ┆ 24     │
│ Demian Maia           ┆ 24     │
│ Andrei Arlovski       ┆ 24     │
│ …                     ┆ …      │
│ Diana Belbita         ┆ 1      │
│ Bea Malecki           ┆ 1      │
│ Anthony Christodoulou ┆ 1      │
│ Branden Lee Hinkle    ┆ 1      │
└───────────────────────┴────────┘
shape: (1_987, 2)
┌──────────────────┬────────┐
│ B_fighter        ┆ counts │
│ ---              ┆ ---    │
│ str              ┆ u32    │
╞══════════════════╪════════╡
│ Jeremy Stephens  ┆ 21     │
│ Charles Oliveira ┆ 19     │
│ Nik Lentz        ┆ 15     │
│ Rafael Dos Anjos ┆ 14     │
│ …                ┆ …      │
│ Tony Petarra     ┆ 1      │
│ Justin Frazier   ┆ 1      │
│ Hernani Perpetuo ┆ 1      │
│ Ednaldo Oliveira ┆ 1      │
└──────────────────

In [16]:
df.shape

(6012, 144)